### DISCUSSION TITLE
Multiple json+ld structured data for SEO

### DISCUSSION THREAD

**Discussion by telestrial** [vercel/next.js#44836] [Help] [answered]
Hello,

My goal here is to have two different application/json+ld scripts in the head when I'm on a product page. 1 organizational that sits on all pages and 1 for individual product pages. I've simplified things just for the sake of the discussion. [Here](https://codesandbox.io/p/sandbox/goofy-julien-vqffs6?selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A24%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A24%7D%5D&file=%2Fpackage.json) is a code sandbox that lays out a couple different strategies I've tried that have not worked.

1. Components representing the script tag inserted between head on _document and the product page. (GlobalSEO and ProductSEO)
2. Components wrapped in Head placed in _document and product outside the head. (WrappedGlobalSEO and WrappedProductSEO).
3. bare script tags in the head of _document and product

Interestingly, it appears the only way to get product page to have the product schema is to actually remove the global schema json altogether.

Another tactic might be to create some kind of super component that produces a json+ld array, 0th position having the organizational data and then conditionally adding another index for the product data if those props exist, but that would be a fair amount of code repetition. Is there a better way?

Some threads I've been looking over that may or may not be related: [#22729](https://github.com/vercel/next.js/issues/22729), [#11012](https://github.com/vercel/next.js/issues/11012), and [#9126](https://github.com/vercel/next.js/issues/9126). This [comment](https://github.com/vercel/next.js/issues/22729#issuecomment-974516530), in particular, seemed to suggest it was completely possible, but I haven't had luck using that method (third on the list above).

**Comment by icyJoseph** [comment, marked as answer]
You could define a data getter at page level, and then from within _app collect the page data, if any, and merge that with some baseline/defaults.

```tsx
import type { AppProps } from "next/app";

type AnyObject = Record<string, unknown>;

type Data = {'@content': string; '@type': string; name: string; pid: number}

export type MyPage<P = AnyObject, IP = AnyObject> = NextPage<P, IP> & {
    getJSONStructuredData?: (pageProps: P) => Array<Data>;
};

type MyAppProps = AppProps & {
    Component: MyPage;
}

const MyApp = ({Component, pageProps, initialState}: MyAppProps) =>  {
    return (
        <>   
                <InjectSEO pageOwnData={Component?.getJSONStructuredData?.(pageProps)} />
                <Component {...pageProps} />
        </>
    );
}
```

And then each page can opt-in to implementing `getJSONStructruredData`:

```tsx
const Product: MyPage<ProductProps> = () => <div>P</div>

// props here is typed as ProductProps
Product. getJSONStructuredData = (props) => { return [ /* something */] } 
```

Pretty much the way shared layouts work in Next.js pages directory: https://nextjs.org/docs/basic-features/layouts

The InjectSEO component takes care of filling in defaults ~

**Reply by telestrial** [reply]
Does

```
export type Page<P = AnyObject, IP = AnyObject> = NextPage<P, IP> & {
    getJSONStructuredData?: (pageProps: P) => Array<Data>;
};
```
need to be MyPage ? I'm not being critical I'm just trying to make sure I understand the example.

**Reply by icyJoseph** [reply]
AH yeah, my bad, it was very early when I wrote that...

**Reply by telestrial** [reply]
No problem! I'm a junior dev so I'm like barely hanging with you here. However: I implemented this and it works really well. This is a way to condense two (or three or infinity) json+ld objects that could be page specific along with other more general structured data you may want to include on a page. The only thing I changed was returning an object instead of an array. I then collect these objects in the InjectSEO component to compile a json+ld array. This reads well and passes schema.org validation. Given that Next seems to do some "magic" in the Head to prevent having two separate tags and Google/others may not even pay attention to a second tag, this is the good stuff.

P.S. In the second snippet, it's also `getJSONStructuredData` instead of `getJSONStructruredData`, just in case anyone else finds their way here.

Thank you so much for your help. This is great.
